src/drivers/fluid_waveout.c

```diff
@@ -0,0 +1,359 @@
+                    device = n;
+                    break;
+                }
+            }
+        }
+    }
+
+    if(dev_name)
+    {
+        FLUID_FREE(dev_name);
+    }
+
+    /* Create thread which processes re-adding SYSEX buffers */
+    dev->hThread = CreateThread(
+                       NULL,
+                       0,
+                       (LPTHREAD_START_ROUTINE)
+                       fluid_waveout_synth_thread,
+                       dev,
+                       0,
+                       &dev->dwThread);
+
+    if(dev->hThread == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to create waveOut thread");
+        HeapFree(GetProcessHeap(), 0, dev);
+        return NULL;
+    }
+
+    errCode = waveOutOpen(&dev->hWaveOut,
+                          device,
+                          &wfx,
+                          (DWORD_PTR)dev->dwThread,
+                          0,
+                          CALLBACK_THREAD);
+
+    if (errCode != MMSYSERR_NOERROR)
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to open waveOut device");
+
+        /* Close the thread */
+        PostThreadMessage(dev->dwThread, WM_CLOSE, 0, 0);
+        WaitForSingleObject(dev->hThread, INFINITE);
+        CloseHandle(dev->hThread);
+
+        HeapFree(GetProcessHeap(), 0, dev);
+        return NULL;
+    }
+
+    /* Get pointer to sound buffer memory */
+    ptrBuffer = (LPSTR)(dev+1);
+
+    /* Setup the sample buffers */
+    for (i = 0; i < NB_SOUND_BUFFERS; i++)
+    {
+        /* Clear the sample buffer */
+        memset(ptrBuffer, 0, lenBuffer);
+
+        /* Clear descriptor buffer */
+        memset(dev->waveHeader+i, 0, sizeof(WAVEHDR));
+
+        /* Compile descriptor buffer */
+        dev->waveHeader[i].lpData         = ptrBuffer;
+        dev->waveHeader[i].dwBufferLength = lenBuffer;
+
+        waveOutPrepareHeader(dev->hWaveOut, &dev->waveHeader[i], sizeof(WAVEHDR));
+
+        ptrBuffer += lenBuffer;
+    }
+
+    /* Play the sample buffers */
+    for (i = 0; i < NB_SOUND_BUFFERS; i++)
+    {
+        waveOutWrite(dev->hWaveOut, &dev->waveHeader[i], sizeof(WAVEHDR));
+    }
+
+    return (fluid_audio_driver_t *) dev;
+}
+
+
+void delete_fluid_waveout_audio_driver(fluid_audio_driver_t *d)
+{
+    int i;
+
+    fluid_waveout_audio_driver_t *dev = (fluid_waveout_audio_driver_t *) d;
+    fluid_return_if_fail(dev != NULL);
+
+    /* release all the allocated resources */
+    if (dev->hWaveOut != NULL)
+    {
+        waveOutReset(dev->hWaveOut);
+        waveOutClose(dev->hWaveOut);
+
+        /* Release the sample buffers */
+        for (i = 0; i < NB_SOUND_BUFFERS; i++)
+        {
+            waveOutUnprepareHeader(dev->hWaveOut, &dev->waveHeader[i], sizeof(WAVEHDR));
+        }
+    }
+
+    if (dev->hThread != NULL)
+    {
+        PostThreadMessage(dev->dwThread, WM_CLOSE, 0, 0);
+        WaitForSingleObject(dev->hThread, INFINITE);
+
+        CloseHandle(dev->hThread);
```

Comment:
Why do you quit the thread after waveOut has been closed? This seems wrong. The thread could be preempted during the call to `fluid_synth_write_*()`. During that time, waveOut could be closed. When the thread wakes up again, it will call a disposed waveOut instance, which is likely to crash.

Reply:
I suppose that such call to waveOutWrite will return MMSYSERR_INVALHANDLE.
If the thread is closed before the waveOut handle, I'm a bit feared about what it could happen inside the waveOut functions, because I have no control on them. Perhaps, we could call waveOutReset, close the thread and finally call waveOutClose, probably it should be less risky.

Reply:
> If the thread is closed before the waveOut handle, I'm a bit feared about what it could happen inside the waveOut functions

I guess it would just keep filling the message queue. Should be unproblematic. WinMidi does it the same way.

> Perhaps, we could call waveOutReset, close the thread and finally call waveOutClose

This should be well defined, but could result in a short sequence of audible sound. Think of:

```c
waveOutReset()
// playback stopped, but thread is still busy inside `fluid_synth_write_*()`
waveOutWrite()
// thread finishes
waveOutUnprepareHeader()
waveOutClose()
```